### ISSUE TITLE
Implement IE Specific Desired Capabilities

### ISSUE THREAD

**Issue by titusfortner** [SeleniumHQ/selenium#1073]
IE Bridge [fails to initialize when using an unknown parameter](https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/ie/bridge.rb#L49)

And most of the [IE specific capabilities](https://github.com/SeleniumHQ/selenium/wiki/DesiredCapabilities#ie-specific) are not currently listed.

I'll eventually get to this, but if any lurking Windows Rubyists want something to work on before that... :)

**Comment by titusfortner** [comment]
Also need to verify if creating driver with this is properly updating native_events:
```
Selenium::WebDriver.for(:ie, :native_events => false)
```
I thought that caps as stored [in the ie bridge](https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/ie/bridge.rb#L58) was supposed to still be snake case symbol and not camel case string, but I need to fire up a Windows VM to verify.

**Comment by trabulmonkee** [comment]
I am seeing different behaviors with click actions for image links specifically.  I see 2 requests to the app server being made, when I expect only one.  Behavior is different between whether I set the native_events via a Selenium::Capabilities or via Selenium::WebDriver.for(:ie, :native_events => false)

**Comment by trabulmonkee** [comment]
Additional Information (Information overload)

WebDriver Version 2.46.2
IEDriverServer.exe 2.48.0.0
IE Version 11.0.9600.18059 (Update Versions 11.0.24 KB3093983)

with no desired_capabilities as :nativeEvents
------------
```ruby
driver = Selenium::WebDriver.new(:ie, :nativeEvents => false)
# error thrown - unknown option: {:nativeEvents=>false}
```

```ruby
driver = Selenium::WebDriver.new(:ie, :nativeEvents => true)
# error thrown - unknown option: {:nativeEvents=>true}
```

with no desired_capabilities as :native_events
------------
```ruby
driver = Selenium::WebDriver.new(:ie, native_events => false)
# fast performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x32bfcf8 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>false**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>false, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", "requireWindowFocus"=>false, "unexpectedAlertBehaviour"=>"dismiss"}


```ruby
driver = Selenium::WebDriver.new(:ie, native_events => true)
# slow performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x32fc190 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>true, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", "requireWindowFocus"=>false, "unexpectedAlertBehaviour"=>"dismiss"}

with desired_capabilities as :nativeEvents
------------
```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :nativeEvents => false
)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
# slow performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x30fee80 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>true, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", "requireWindowFocus"=>false, "unexpectedAlertBehaviour"=>"dismiss"}

```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :nativeEvents => true
)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
# slow performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x326ab28 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>true, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", "requireWindowFocus"=>false, "unexpectedAlertBehaviour"=>"dismiss"}

with desired_capabilities as :native_events
------------
```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :native_events => false
)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
# slow performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x30c86e8 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>true, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", "requireWindowFocus"=>false, "unexpectedAlertBehaviour"=>"dismiss"}

```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :native_events => true
)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
# slow performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x319f0b0 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>true, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", "requireWindowFocus"=>false, "unexpectedAlertBehaviour"=>"dismiss"}

adding in requireWindowFocus to desired capabilities as :nativeEvents
------------
```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :nativeEvents => false,
  :requireWindowFocus => true
)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
# fast performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x325e510 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>false, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", **"requireWindowFocus"=>true**, "unexpectedAlertBehaviour"=>"dismiss"}

```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :nativeEvents => true,
  :requireWindowFocus => true
)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
# fast performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x30fd578 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>false, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", **"requireWindowFocus"=>true**, "unexpectedAlertBehaviour"=>"dismiss"}

adding in requireWindowFocus to desired capabilities as :native_events
------------
```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :native_events => false,
  :requireWindowFocus => false
)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
# slow performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x33ae120 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>true, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", **"requireWindowFocus"=>false**, "unexpectedAlertBehaviour"=>"dismiss"}

```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :native_events => true,
  :requireWindowFocus => false
)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
# slow performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x31decf8 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>true, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", **"requireWindowFocus"=>false**, "unexpectedAlertBehaviour"=>"dismiss"}

```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :native_events => false,
  :requireWindowFocus => true
)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
# fast performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x31fe120 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>false, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", **"requireWindowFocus"=>true**, "unexpectedAlertBehaviour"=>"dismiss"}

```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :native_events => true,
  :requireWindowFocus => true
)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
# fast performance
```
**Produces:**
Selenium::WebDriver::Remote::Capabilities:0x31bee08 @capabilities={:browser_name=>"internet explorer", :version=>"11", :platform=>:windows, :javascript_enabled=>true, :css_selectors_enabled=>true, :takes_screenshot=>true, **:native_events=>true**, :rotatable=>nil, :firefox_profile=>nil, :proxy=>nil, "browserAttachTimeout"=>0, "elementScrollBehavior"=>0, "enableElementCacheCleanup"=>true, "enablePersistentHover"=>false, "handlesAlerts"=>true, "ie.browserCommandLineSwitches"=>"", "ie.ensureCleanSession"=>false, "ie.forceCreateProcessApi"=>false, "ie.usePerProcessProxy"=>false, "ignoreProtectedModeSettings"=>false, "ignoreZoomSetting"=>false, "initialBrowserUrl"=>"http://localhost:5555/", "pageLoadStrategy"=>"normal", **"requireWindowFocus"=>true**, "unexpectedAlertBehaviour"=>"dismiss"}

**Comment by titusfortner** [comment]
That's interesting that NativeEvents doesn't properly toggle but requireWindowFocus does. I'm guessing since there is explicit handling of native events and not require window focus that something is causing issues.
It's also disturbing that this doesn't work:
```
caps = Selenium::WebDriver::Remote::Capabilities.ie(:native_events => false)
driver = Selenium::WebDriver.new(:ie, :desired_capabilities => caps)
```
 Thanks for noting the performance distinctions as well. This will be fun to dig into.

**Comment by trabulmonkee** [comment]
Indeed.  Changes in behavior were also seen with the double click issues that I mentioned on 10-22 above.  I have corrected a typo with my 11-20-2015 post, changed .new to .for (.new presents other problems)

Another interesting item to note is that requireWindowFocus is overwriting the native_events value, when 
```ruby
caps = Selenium::WebDriver::Remote::Capabilities.ie(
  :native_events => false,
  :requireWindowFocus => true
)
```
Perhaps @jimevans may be able to help point us in the right direction.
For what it is worth, I have attached a xUnit style (minitest) test suite used
[native_events_tester.txt](https://github.com/SeleniumHQ/selenium/files/41739/native_events_tester.txt)

**Comment by titusfortner** [comment]
Of interest, if you set $DEBUG = true, you can see the exact json that gets sent to the session endpoint. I'm pretty sure this is a ruby binding issue in how we are translating our code to that json. First step is verifying what the right json is, then making sure all the different ways we support coding it results in that json. Then we can figure out why you are seeing a performance difference.

I got a Windows virtual machine up and running this past weekend to work on Edge implementation things. I should be able to dig into this soon.

**Comment by trabulmonkee** [comment]
Created a repo location to test this further.  Added debug logs and updated the native_events_tester test suite.  All information available here: http://trabulmonkee.github.io/se-repos/ie-native-events/

**Comment by abaird** [comment]
@trabulmonkee - I have your tests implemented in my forkk here: https://github.com/abaird/selenium/commits/issue_1073 - LMK if you have questions

**Comment by abaird** [comment]
### [For the case where we are making our own Capabilities object and passing it to Selenium::WebDriver.for with :native_events => false](https://github.com/abaird/selenium/blob/issue_1073/rb/spec/integration/selenium/webdriver/capabilities_spec.rb#L66-L71)

I figured out the bug.

**TL;DR** - when the JSON is generated to make a wire call, [Selenium::Webdriver::Remote::Capabilities#as_json](https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/remote/capabilities.rb#L239-L260) is called to convert itself into a JSON string. At this point, the Capabilities object has 2 keys in @capabilities - `:native_events` and `'nativeEvents'`, and #as_json converts the symbol into camel case. Thus, whichever key is processed last keeps it's value.

********

The first problem happens in [Selenium::WebDriver::IE::Bridge.new](https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/ie/bridge.rb#L58). For some reason, Jari makes a new key in the capabilities object called `'nativeEvents'`. This sets up the problem because the Capabilities object that is passed to the Remote::Bridge now has a key `:native_events` and a key `'nativeEvents'`. This is done because it's valid to pass in `:native_events` as a key for options and we need to make sure it is properly set. (But I don't really understand why we do that since we will create an IE Capabilities object (with :native_events => true) by default anyway. Why can't we just use this object to pass the `:native_events` key? - I digress) 

Eventually, that object (with the two keys) gets passed off to [Selenium::Webdriver::Remote::Bridge#create_session](https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/remote/bridge.rb#L87). This Capabilities object is they value of the `:desiredCapabilities` key in the command hash that gets JSONified in the wire call. In [Selenium::WebDriver::Remote::Http::Common#call](https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/remote/http/common.rb#L46), this object is sent to `JSON.generate`. It took me a while to figure out what happens here. JSON translates the `:desiredCapabilities` key into a string, then it has to translate the value of this key - which is an object. In order to do this, `JSON.generate` calls the #as_json method on that object - which is defined in [Selenium::WebDriver::Remote::Capabilities#as_json](https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/remote/capabilities.rb#L239-L260). If you look at this method, you will see that it eaches over capabilities, mapping any symbols it finds using the #camel_case method that's provided in this class.

Thus, as the object is being JSONified, it converts `'nativeEvents'` (the key supplied ...IE::Bridge.new) into `'nativeEvents'` and it converts `:native_events` (the key we originally supplied) into `'nativeEvents'`! At this point, these are two of the same keys in a hash and the last one converted wins. 
```ruby
irb(main):007:0> JSON.generate(:native_events => true, :native_events => false)
(irb):7: warning: duplicated key at line 7 ignored: :native_events
=> "{\"native_events\":false}"
```
In this case, it's the `'nativeEvents'` key that's supplied by IE::Bridge.new - thus squashing the key we meant to send.

Then, when the capabilities object is returned from the wire call, it calls the [...Capabilities#json_create](https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/remote/capabilities.rb#L161-L180) method using the response value. When the values come back from the wire call, they are all camel case. However, #json_create does the translation back to snake_case for us.

Thus, the round trip converts the `'nativeEvents'` key that is made up by ...IE::Bridge.new and shoves it's value into the native_events key returning from the wire call.

I don't know what the solution to this is. I think that we should eliminate the `caps['nativeEvents'] = native_events` line in [...IE::Bridge.new](https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/ie/bridge.rb#L58) but I'm sure that will introduce new bugs. Hopefully by going through the rest of these cases we can hammer all of that out.

**Comment by titusfortner** [comment]
Yes, this is what I remember from our coding session in December. 

Ruby bindings are a bit weird in that they specify some desired capabilities explicitly (using snake case symbols), and then default to letting everything else be their typical camel case strings, which then sometimes results in weird conversions happening. After 3.0 gets released, we're going to revamp the Ruby capabilities classes to be more explicit about everything (which I think should be easier as we move to w3c compliance). 

In the meantime, I believe these are the (only) two ways that need to work, and I think it is just a line or two of code to make it work:
```
caps = Capabilities.new(native_events: true)
caps.native_events # => true
```
```
caps = Capabilities.new
caps[:native_events] = true
caps.native_events # => true
```
Let's discuss on slack how you guys are doing the next PR. :)

**Comment by titusfortner** [comment]
This should be fixed by - https://github.com/SeleniumHQ/selenium/commit/9c3e0c
@trabulmonkee - let me know if there's something else that needs to get fixed here.
